### PR TITLE
[IMP] mail: Attachments in the BoM are not available in the product.te…

### PR DESCRIPTION
**Pull request by anr-odoo**
Add an attachment on a product on the BoM it will be available on the
product.product, as well as in the product.template.

Task : https://www.odoo.com/web#id=1982093&action=327&model=project.task&view_type=form&menu_id=4720
pad : https://pad.odoo.com/p/r.1318b9a1752012ba58d7c2666713db11 

task id-1982093

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product/models/product_template.py [hunk 1]
```diff
@@ -1108,3 +1108,20 @@ def get_import_templates(self):
             'label': _('Import Template for Products'),
             'template': '/product/static/xls/product_template.xls'
         }]
+
+    @api.multi
+    def _compute_message_attachment_count(self):
+        super(ProductTemplate, self)._compute_message_attachment_count()
```
Comment (sswapnesh) [inline addons/product/models/product_template.py]:
```suggestion
        super()._compute_message_attachment_count()
```
(Reference from https://github.com/odoo/odoo/pull/33975#discussion_r291676373)

File: addons/mail/models/mail_thread.py [hunk 2]
```diff
@@ -2747,3 +2747,13 @@ def _get_mail_redirect_suggested_company(self):
         if 'company_id' in self:
             return self.company_id
         return False
+
+    @api.model
+    def fetch_attachments(self, res_id):
```
Comment (sswapnesh) [inline addons/mail/models/mail_thread.py]: Would be great to have a way to extend this (domain and fields)

File: addons/product/models/product_template.py [hunk 3]
```diff
@@ -1108,3 +1108,20 @@ def get_import_templates(self):
...
+        if record:
+            record.message_attachment_count = record.message_attachment_count + record.product_variant_id.message_attachment_count
+
+    @api.model
+    def fetch_attachments(self, res_id):
+        res = super(ProductTemplate, self).fetch_attachments(res_id)
+        template = self.browse(res_id)
+        if template:
+            if len(template.product_variant_ids) == 1:
+                res += self.env['product.product'].fetch_attachments(template.product_variant_ids[0].id)
```
Comment (fmdl) [inline addons/product/models/product_template.py]: `[0]` is not mandatory

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -2746,3 +2746,12 @@ def _get_mail_redirect_suggested_company(self):
         if 'company_id' in self:
             return self.company_id
         return False
+
+    def read_attachments(self):
```
Comment (sle-odoo) [inline addons/mail/models/mail_thread.py]: self.ensure_one if we'r e using self.id below

File: addons/mail/models/mail_thread.py [hunk 5]
```diff
@@ -2746,3 +2746,12 @@ def _get_mail_redirect_suggested_company(self):
...
+        domain = [
+            ('res_id', '=', self.id),
+            ('res_model', '=', self._name)
+        ]
+        fields = ['id', 'name', 'datas_fname', 'mimetype']
```
Comment (sle-odoo) [inline addons/mail/models/mail_thread.py]: datas_fname was not passed in the original js?
```
fields: ['id', 'name', 'mimetype'],
```

File: addons/mail/models/mail_thread.py [hunk 6]
```diff
@@ -2746,3 +2746,12 @@ def _get_mail_redirect_suggested_company(self):
...
+        res = self.env['ir.attachment'].search_read(domain, fields)
```
Comment (sle-odoo) [inline addons/mail/models/mail_thread.py]: return self.env['ir.attachment'].search_read(domain, fields)

File: addons/product/models/product_template.py [hunk 7]
```diff
@@ -1108,3 +1108,16 @@ def get_import_templates(self):
             'label': _('Import Template for Products'),
             'template': '/product/static/xls/product_template.xls'
         }]
+
+    @api.multi
+    def _compute_message_attachment_count(self):
+        super(ProductTemplate, self)._compute_message_attachment_count()
+        record = self.filtered(lambda x: len(x.product_variant_ids) == 1)
```
Comment (odony) [inline addons/product/models/product_template.py]:
```suggestion
        records = self.filtered(lambda x: len(x.product_variant_ids) == 1)
```
Rename the variable here, as this is not always a singleton. And you need to handle those multiple records properly, as the next lines will crash.
(Easy to test: add `message_attachment_count` in the kanban view)

File: addons/mail/models/mail_thread.py (hunk 5 above)
Comment (anr-odoo) [reply addons/mail/models/mail_thread.py]: done

File: addons/mail/models/mail_thread.py (hunk 6 above)
Comment (anr-odoo) [reply addons/mail/models/mail_thread.py]: done

File: addons/product/models/product_template.py (hunk 3 above)
Comment (anr-odoo) [reply addons/product/models/product_template.py]: done

**Pull request closed without merging**